File: src/lib/utilities/bill-storage.ts
/**
 * Standardized bill storage utilities
 *
 * All utility bills are stored in:
 *   data/bills/{address-slug}/{provider}_{YYYY-MM}.pdf
 *
 * Example:
 *   data/bills/3606-appling-way/duke-energy_2026-01.pdf
 *   data/bills/118-king-arthur-ct/enbridge-gas_2026-01.pdf
 */

import * as fs from "fs";
import * as path from "path";

const DATA_DIR = path.join(process.cwd(), "data", "bills");

/**
 * Convert an address to a URL-safe slug
 * "3606 APPLING WAY, DURHAM NC 27703" -> "3606-appling-way"
 * "310B HOWARD ST" -> "310b-howard-st"
 * "310 HOWARD ST B" -> "310b-howard-st"
 */
export function addressToSlug(address: string): string {
  // Take just the street address (before city/state/zip)
  let streetAddress = address.split(",")[0].trim();

  // Remove unit designations for the folder (keep it simple)
  streetAddress = streetAddress
    .replace(/\s+(UNIT|APT|STE|SUITE|#)\s*\w*/gi, "")
    .trim();

  // Handle trailing unit letters (like "310 HOWARD ST B" -> "310B HOWARD ST")
  // This normalizes addresses with unit letters at the end
  const trailingUnitMatch = streetAddress.match(/^(\d+)\s+(.+?)\s+([A-Za-z])$/);
  if (trailingUnitMatch) {
    const num = trailingUnitMatch[1];
    const street = trailingUnitMatch[2];
    const unit = trailingUnitMatch[3].toUpperCase();
    streetAddress = `${num}${unit} ${street}`;
  }

  // Convert to lowercase, replace spaces/special chars with dashes
  return streetAddress
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, ""); // trim leading/trailing dashes
}

/**
 * Convert a provider name to a URL-safe slug
 * "Duke Energy" -> "duke-energy"
 * "Enbridge Gas" -> "enbridge-gas"
 */
export function providerToSlug(provider: string): string {
  return provider
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

/**
 * Format a billing period from a date
 * Returns "YYYY-MM" format
 */
export function formatBillingPeriod(date: Date): string {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, "0");
  return `${year}-${month}`;
}

/**
 * Get the directory path for a property's bills
 */
export function getBillDirectory(address: string): string {
  const slug = addressToSlug(address);
  return path.join(DATA_DIR, slug);
}

/**
 * Get the full path for a bill PDF
 *
 * @param address - The service address (e.g., "3606 APPLING WAY, DURHAM NC")
 * @param provider - The utility provider (e.g., "Duke Energy")
 * @param billingDate - The billing date (used to determine YYYY-MM)
 * @returns Full path like "data/bills/3606-appling-way/duke-energy_2026-01.pdf"
 */
export function getBillPath(
  address: string,
  provider: string,
  billingDate: Date
): string {
  const dir = getBillDirectory(address);
  const providerSlug = providerToSlug(provider);
  const period = formatBillingPeriod(billingDate);
  return path.join(dir, `${providerSlug}_${period}.pdf`);
}

/**
 * Ensure the bill directory exists and return the path
 */
export function ensureBillDirectory(address: string): string {
  const dir = getBillDirectory(address);
  fs.mkdirSync(dir, { recursive: true });
  return dir;
}

/**
 * Save a bill PDF to the standardized location
 *
 * @param address - The service address
 * @param provider - The utility provider
 * @param billingDate - The billing date
 * @param pdfBuffer - The PDF content as a Buffer
 * @returns The path where the file was saved
 */
export function saveBillPdf(
  address: string,
  provider: string,
  billingDate: Date,
  pdfBuffer: Buffer
): string {
  ensureBillDirectory(address);
  const filePath = getBillPath(address, provider, billingDate);

  // If file already exists, add a timestamp suffix to avoid overwriting
  let finalPath = filePath;
  if (fs.existsSync(filePath)) {
    const timestamp = Date.now();
    const ext = path.extname(filePath);
    const base = filePath.slice(0, -ext.length);
    finalPath = `${base}_${timestamp}${ext}`;
  }

  fs.writeFileSync(finalPath, pdfBuffer);
  return finalPath;
}

/**
 * Copy an existing bill PDF to the standardized location
 */
export function copyBillToStandardLocation(
  sourcePath: string,
  address: string,
  provider: string,
  billingDate: Date
): string {
  if (!fs.existsSync(sourcePath)) {
    throw new Error(`Source file not found: ${sourcePath}`);
  }

  ensureBillDirectory(address);
  const destPath = getBillPath(address, provider, billingDate);

  // Don't overwrite if same content
  if (fs.existsSync(destPath)) {
    const sourceSize = fs.statSync(sourcePath).size;
    const destSize = fs.statSync(destPath).size;
    if (sourceSize === destSize) {
      return destPath; // Already exists with same size
    }
    // Different file, add timestamp
    const timestamp = Date.now();
    const ext = path.extname(destPath);
    const base = destPath.slice(0, -ext.length);
    const newPath = `${base}_${timestamp}${ext}`;
    fs.copyFileSync(sourcePath, newPath);
    return newPath;
  }

  fs.copyFileSync(sourcePath, destPath);
  return destPath;
}

/**
 * List all bills for a property
 */
export function listBillsForProperty(address: string): string[] {
  const dir = getBillDirectory(address);
  if (!fs.existsSync(dir)) {
    return [];
  }
  return fs
    .readdirSync(dir)
    .filter((f) => f.endsWith(".pdf"))
    .map((f) => path.join(dir, f));
}

/**
 * Get all bill directories
 */
export function listAllBillDirectories(): string[] {
  if (!fs.existsSync(DATA_DIR)) {
    return [];
  }
  return fs
    .readdirSync(DATA_DIR)
    .filter((f) => fs.statSync(path.join(DATA_DIR, f)).isDirectory())
    .map((f) => path.join(DATA_DIR, f));
}

File: scripts/lib/bill_storage.py
"""
Standardized bill storage utilities for Python scrapers.

All utility bills are stored in:
  data/bills/{address-slug}/{provider}_{YYYY-MM}.pdf

Example:
  data/bills/3606-appling-way/duke-energy_2026-01.pdf
  data/bills/118-king-arthur-ct/enbridge-gas_2026-01.pdf
"""

import os
import re
from datetime import datetime
from pathlib import Path
from typing import Optional


def get_project_root() -> Path:
    """Get the project root directory."""
    # Go up from scripts/lib to project root
    return Path(__file__).parent.parent.parent


def get_bills_directory() -> Path:
    """Get the base bills directory."""
    return get_project_root() / "data" / "bills"


def address_to_slug(address: str) -> str:
    """
    Convert an address to a URL-safe slug.
    "3606 APPLING WAY, DURHAM NC 27703" -> "3606-appling-way"
    "310B HOWARD ST" -> "310b-howard-st"
    "310 HOWARD ST B" -> "310b-howard-st"
    """
    # Take just the street address (before city/state/zip)
    street_address = address.split(",")[0].strip()

    # Remove unit designations for the folder (keep it simple)
    street_address = re.sub(r'\s+(UNIT|APT|STE|SUITE|#)\s*\w*', '', street_address, flags=re.IGNORECASE)
    street_address = street_address.strip()

    # Handle trailing unit letters (like "310 HOWARD ST B" -> "310B HOWARD ST")
    # This normalizes addresses with unit letters at the end
    trailing_unit_match = re.match(r'^(\d+)\s+(.+?)\s+([A-Z])$', street_address, re.IGNORECASE)
    if trailing_unit_match:
        num = trailing_unit_match.group(1)
        street = trailing_unit_match.group(2)
        unit = trailing_unit_match.group(3).upper()
        street_address = f"{num}{unit} {street}"

    # Convert to lowercase, replace spaces/special chars with dashes
    slug = re.sub(r'[^a-z0-9]+', '-', street_address.lower())
    slug = slug.strip('-')  # trim leading/trailing dashes

    return slug


def provider_to_slug(provider: str) -> str:
    """
    Convert a provider name to a URL-safe slug.
    "Duke Energy" -> "duke-energy"
    """
    slug = re.sub(r'[^a-z0-9]+', '-', provider.lower())
    return slug.strip('-')


def format_billing_period(date: datetime) -> str:
    """Format a billing period from a date. Returns "YYYY-MM" format."""
    return date.strftime("%Y-%m")


def get_bill_directory(address: str) -> Path:
    """Get the directory path for a property's bills."""
    slug = address_to_slug(address)
    return get_bills_directory() / slug


def get_bill_path(address: str, provider: str, billing_date: datetime) -> Path:
    """
    Get the full path for a bill PDF.

    Args:
        address: The service address (e.g., "3606 APPLING WAY, DURHAM NC")
        provider: The utility provider (e.g., "Duke Energy")
        billing_date: The billing date (used to determine YYYY-MM)

    Returns:
        Full path like "data/bills/3606-appling-way/duke-energy_2026-01.pdf"
    """
    dir_path = get_bill_directory(address)
    provider_slug = provider_to_slug(provider)
    period = format_billing_period(billing_date)
    return dir_path / f"{provider_slug}_{period}.pdf"


def ensure_bill_directory(address: str) -> Path:
    """Ensure the bill directory exists and return the path."""
    dir_path = get_bill_directory(address)
    dir_path.mkdir(parents=True, exist_ok=True)
    return dir_path


def save_bill_pdf(
    address: str,
    provider: str,
    billing_date: datetime,
    pdf_content: bytes
) -> str:
    """
    Save a bill PDF to the standardized location.

    Args:
        address: The service address
        provider: The utility provider
        billing_date: The billing date
        pdf_content: The PDF content as bytes

    Returns:
        The path where the file was saved
    """
    ensure_bill_directory(address)
    file_path = get_bill_path(address, provider, billing_date)

    # If file already exists with same size, skip
    if file_path.exists():
        existing_size = file_path.stat().st_size
        if existing_size == len(pdf_content):
            print(f"  Bill already exists: {file_path}")
            return str(file_path)

        # Different content, add timestamp suffix
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        new_path = file_path.with_stem(f"{file_path.stem}_{timestamp}")
        file_path = new_path

    with open(file_path, 'wb') as f:
        f.write(pdf_content)

    print(f"  Saved bill: {file_path}")
    return str(file_path)


def copy_bill_to_standard_location(
    source_path: str,
    address: str,
    provider: str,
    billing_date: datetime
) -> str:
    """Copy an existing bill PDF to the standardized location."""
    source = Path(source_path)
    if not source.exists():
        raise FileNotFoundError(f"Source file not found: {source_path}")

    ensure_bill_directory(address)
    dest_path = get_bill_path(address, provider, billing_date)

    # Check if already exists with same content
    if dest_path.exists():
        source_size = source.stat().st_size
        dest_size = dest_path.stat().st_size
        if source_size == dest_size:
            return str(dest_path)  # Already exists

        # Different content, add timestamp
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        dest_path = dest_path.with_stem(f"{dest_path.stem}_{timestamp}")

    import shutil
    shutil.copy2(source_path, dest_path)
    return str(dest_path)
